File: AgoraChat/Code/Controllers/Groups/ACDGroupMemberAllViewController.m
//
//  ACDGroupMemberAllViewController.m
//  ChatDemo-UI3.0
//
//  Created by liang on 2021/10/29.
//  Copyright © 2021 easemob. All rights reserved.
//

#import "ACDGroupMemberAllViewController.h"
#import "AgoraGroupOccupantsViewController.h"
#import "UIViewController+HUD.h"
#import "AgoraNotificationNames.h"
#import "ACDContainerSearchTableViewController+GroupMemberList.h"
#import "ACDContactCell.h"
#import "NSArray+AgoraSortContacts.h"
#import "AgoraUserModel.h"

@interface ACDGroupMemberAllViewController ()

@property (nonatomic, strong) NSString *groupId;
@property (nonatomic, strong) AgoraChatGroup *group;
@property (nonatomic, strong) NSString *cursor;

@end

@implementation ACDGroupMemberAllViewController

- (instancetype)initWithGroup:(AgoraChatGroup *)aGroup
{
    self = [super init];
    if (self) {
        self.group = aGroup;
        self.groupId = self.group.groupId;
        
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(updateGroupMemberWithNotification:) name:KACD_REFRESH_GROUP_MEMBER object:nil];
    }
    
    return self;
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self useRefresh];
    [self tableViewDidTriggerHeaderRefresh];
}

- (void)updateUIWithResultList:(NSArray *)sourceList IsHeader:(BOOL)isHeader {
    
    if (isHeader) {
        [self.members removeAllObjects];
        [self.members addObject:self.group.owner];
        [self.members addObjectsFromArray:self.group.adminList];
    }

    [self.members addObjectsFromArray:sourceList];
    
    [self sortContacts:self.members];

    dispatch_async(dispatch_get_main_queue(), ^(){
        [self.table reloadData];
    });
}





- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (void)updateUI {
    [self tableViewDidTriggerHeaderRefresh];
}

#pragma mark updateUIWithNotification
- (void)updateGroupMemberWithNotification:(NSNotification *)aNotification {
    NSDictionary *dic = (NSDictionary *)aNotification.object;
    NSString* groupId = dic[kACDGroupId];
    
    if (![self.group.groupId isEqualToString:groupId]) {
        return;
    }
    
    [self tableViewDidTriggerHeaderRefresh];

}



#pragma mark - Table view data source
- (NSInteger)numberOfSectionsInTableView:(UITableView *)table {
    if (self.isSearchState) {
        return 1;
    }
    return  self.sectionTitles.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (self.isSearchState) {
        return self.searchResults.count;
    }
    return ((NSArray *)self.dataArray[section]).count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return [ACDContactCell height];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    ACDContactCell *cell = (ACDContactCell *)[tableView dequeueReusableCellWithIdentifier:[ACDContactCell reuseIdentifier]];
    if (cell == nil) {
        cell = [[ACDContactCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:[ACDContactCell reuseIdentifier]];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
    }
    
    AgoraUserModel *model = nil;
    if (self.isSearchState) {
        model = self.searchResults[indexPath.row];
    }else {
        model = self.dataArray[indexPath.section][indexPath.row];
    }
    
    cell.model =  model;
    if (model.hyphenateId == self.group.owner) {
        cell.detailLabel.text = @"owner";
    } else if([self.group.adminList containsObject:model.hyphenateId]){
        cell.detailLabel.text = @"admin";
    }else {
        cell.detailLabel.text = @"";
    }

    ACD_WS
    cell.tapCellBlock = ^{
        [weakSelf actionSheetWithUserId:model.hyphenateId memberListType:ACDGroupMemberListTypeALL group:weakSelf.group];
    };
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
}

#pragma mark refresh and load more
- (void)didStartRefresh {
    [self tableViewDidTriggerHeaderRefresh];
}

- (void)didStartLoadMore {
    [self tableViewDidTriggerFooterRefresh];
}


#pragma mark - private
- (void)tableViewDidTriggerHeaderRefresh
{
    self.cursor = @"";
    [self fetchMembersWithCursor:self.cursor isHeader:YES];

}

- (void)tableViewDidTriggerFooterRefresh
{
    [self fetchMembersWithCursor:self.cursor isHeader:NO];
}


- (void)fetchMembersWithCursor:(NSString *)aCursor
                      isHeader:(BOOL)aIsHeader
{
    NSInteger pageSize = 50;
        
    ACD_WS

    [[AgoraChatClient sharedClient].groupManager getGroupMemberListFromServerWithId:self.groupId cursor:aCursor pageSize:pageSize completion:^(AgoraChatCursorResult *aResult, AgoraChatError *aError) {
        weakSelf.cursor = aResult.cursor;

        [self endRefresh];
        
        if (!aError) {
            [weakSelf updateUIWithResultList:aResult.list IsHeader:aIsHeader];
        } else {
            [weakSelf showHint:@"Failed to get the group details, please try again later"];
        }
        

        if ([aResult.list count] < pageSize) {
            [weakSelf endLoadMore];
            [weakSelf loadMoreCompleted];
        } else {
            [weakSelf useLoadMore];
        }

        [weakSelf.table reloadData];

    }];
}


@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Groups/ACDGroupMemberAllViewController.swift
//
//  ACDGroupMemberAllViewController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/21.
//

import UIKit

class ACDGroupMemberAllViewController: ACDContainerSearchTableViewController<[AgoraUserModel]> {
    
    private let group: AgoraChatGroup
    private var cursor: String?
    
    init(group: AgoraChatGroup) {
        self.group = group
        super.init(nibName: nil, bundle: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(updateGroupMember(_:)), name: GroupMemberChangedNotification, object: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.table.rowHeight = 54
        self.table.register(UINib(nibName: "ACDContactCell", bundle: nil), forCellReuseIdentifier: "cell")
        
        self.useRefresh()
        self.fetchMembers(isRefresh: true)
    }
    
    func updateUI() {
        self.fetchMembers(isRefresh: true)
    }
    
    @objc private func updateGroupMember(_ notification: Notification) {
        if let groupId = (notification.object as? [String: Any])?["kACDGroupId"] as? String {
            if groupId == self.group.groupId {
                self.updateUI()
            }
        }
    }
    
    private func fetchMembers(isRefresh: Bool) {
        if isRefresh {
            self.cursor = nil
        }
        AgoraChatClient.shared().groupManager?.getGroupMemberListFromServer(withId: self.group.groupId, cursor: self.cursor, pageSize: 20, completion: { result, error in
            self.cursor = result?.cursor
            self.endRefresh()
            if let error = error {
                self.showHint(error.errorDescription)
            } else if let members = result?.list as? [String] {
                if isRefresh {
                    self.members.removeAll()
                    if let owner = self.group.owner {
                        self.members.append(owner)
                    }
                    self.members.append(contentsOf: self.group.adminList)
                }
                self.members.append(contentsOf: members)
                self.sortAndReloadContacts(self.members)
            }
            
            if let list = result?.list, list.count >= 20 {
                self.useLoadMore()
            } else {
                self.endLoadMore()
                self.loadMoreCompleted()
            }
            self.table.reloadData()
        })
    }
    
    override func didStartRefresh() {
        self.fetchMembers(isRefresh: true)
    }
    
    override func didStartLoadMore() {
        self.fetchMembers(isRefresh: false)
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return self.isSearchState ? 1 : self.sectionTitles.count
    }
    
    func sectionIndexTitles(for tableView: UITableView) -> [String]? {
        return self.sectionTitles
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.isSearchState ? self.searchResults.count : self.dataArray[section].count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
        if let cell = cell as? ACDContactCell {
            var model: AgoraUserModel?
            if self.isSearchState {
                model = self.searchResults[indexPath.row] as? AgoraUserModel
            } else {
                model = self.dataArray[indexPath.section][indexPath.row]
            }
            if let model = model {
                cell.model = model
                if model.hyphenateId == self.group.owner {
                    cell.detailLabel.text = "Owner".localized
                } else if self.group.adminList.contains(model.hyphenateId) {
                    cell.detailLabel.text = "Admin".localized
                } else {
                    cell.detailLabel.text = ""
                }
                cell.tapCellHandle = { [unowned self] in
                    self.showActionSheet(userId: model.hyphenateId, showType: .all, group: self.group)
                }
            }
        }
        return cell
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}
